[crd-gen] The generator fails with "ClassCastException" when using SchemaFrom

### Describe the bug

Trying to run the CRD generator in an external minimal codebase fails on any `SchemaFrom` annotation with:

```bash
Fatal error compiling: java.lang.ClassCastException: class io.sundr.model.ClassRef cannot be cast to class java.lang.Class (io.sundr.model.ClassRef is in unnamed module of loader java.net.URLClassLoader @2d2b6960; java.lang.Class is in module java.base of loader 'bootstrap')
```

### Fabric8 Kubernetes Client version

next (development version)

### Steps to reproduce

- Install locally the `crd-generator` from the `master` banch
- clone this repo: https://github.com/andreaTP/schemafrom-classref-repro
- run `mvn clean compile -e`

### Expected behavior

The compilation succeed and the CRD is correctly generated.

### Runtime

other (please specify in additional context)

### Kubernetes API Server version

next (development version)

### Environment

macOS

### Fabric8 Kubernetes Client Logs

```shell
Caused by: java.lang.ClassCastException: class io.sundr.model.ClassRef cannot be cast to class java.lang.Class (io.sundr.model.ClassRef is in unnamed module of loader java.net.URLClassLoader @237f7970; java.lang.Class is in module java.base of loader 'bootstrap')
    at io.fabric8.crd.generator.AbstractJsonSchema$PropertyOrAccessor.lambda$process$0 (AbstractJsonSchema.java:236)
    at java.util.ArrayList.forEach (ArrayList.java:1541)
    at io.fabric8.crd.generator.AbstractJsonSchema$PropertyOrAccessor.process (AbstractJsonSchema.java:212)
    at io.fabric8.crd.generator.AbstractJsonSchema$PropertyFacade.lambda$process$0 (AbstractJsonSchema.java:322)
    at java.util.ArrayList.forEach (ArrayList.java:1541)
    at io.fabric8.crd.generator.AbstractJsonSchema$PropertyFacade.process (AbstractJsonSchema.java:321)
    at io.fabric8.crd.generator.AbstractJsonSchema.internalFromImpl (AbstractJsonSchema.java:148)
    at io.fabric8.crd.generator.AbstractJsonSchema.resolveNestedClass (AbstractJsonSchema.java:503)
    at io.fabric8.crd.generator.AbstractJsonSchema.internalFromImpl (AbstractJsonSchema.java:479)
    at io.fabric8.crd.generator.AbstractJsonSchema.internalFromImpl (AbstractJsonSchema.java:156)
    at io.fabric8.crd.generator.AbstractJsonSchema.internalFrom (AbstractJsonSchema.java:123)
    at io.fabric8.crd.generator.v1beta1.JsonSchema.from (JsonSchema.java:47)
    at io.fabric8.crd.generator.v1beta1.CustomResourceHandler.addDecorators (CustomResourceHandler.java:68)
    at io.fabric8.crd.generator.AbstractCustomResourceHandler.handle (AbstractCustomResourceHandler.java:72)
    at io.fabric8.crd.generator.v1beta1.CustomResourceHandler.handle (CustomResourceHandler.java:99)
    at io.fabric8.crd.generator.CRDGenerator.lambda$detailedGenerate$4 (CRDGenerator.java:154)
    at java.util.HashMap$Values.forEach (HashMap.java:977)
    at io.fabric8.crd.generator.CRDGenerator.detailedGenerate (CRDGenerator.java:154)
    at io.fabric8.crd.generator.apt.CustomResourceAnnotationProcessor.process (CustomResourceAnnotationProcessor.java:54)
    at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor (JavacProcessingEnvironment.java:985)
    at com.sun.tools.javac.processing.JavacProcessingEnvironment$DiscoveredProcessors$ProcessorStateIterator.runContributingProcs (JavacProcessingEnvironment.java:821)
    at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run (JavacProcessingEnvironment.java:1225)
    at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing (JavacProcessingEnvironment.java:1362)
    at com.sun.tools.javac.main.JavaCompiler.processAnnotations (JavaCompiler.java:1258)
    at com.sun.tools.javac.main.JavaCompiler.compile (JavaCompiler.java:936)
    at com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0 (JavacTaskImpl.java:104)
    at com.sun.tools.javac.api.JavacTaskImpl.handleExceptions (JavacTaskImpl.java:147)
    at com.sun.tools.javac.api.JavacTaskImpl.doCall (JavacTaskImpl.java:100)
    at com.sun.tools.javac.api.JavacTaskImpl.call (JavacTaskImpl.java:94)
    at org.codehaus.plexus.compiler.javac.JavaxToolsCompiler.compileInProcess (JavaxToolsCompiler.java:126)
    at org.codehaus.plexus.compiler.javac.JavacCompiler.performCompile (JavacCompiler.java:169)
    at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute (AbstractCompilerMojo.java:785)
    at org.apache.maven.plugin.compiler.CompilerMojo.execute (CompilerMojo.java:129)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:972)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:566)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
```


### Additional context

_No response_

--- a/crd-generator/api/src/main/java/io/fabric8/crd/generator/AbstractJsonSchema.java
+++ b/crd-generator/api/src/main/java/io/fabric8/crd/generator/AbstractJsonSchema.java
@@ -191,7 +191,7 @@ public abstract class AbstractJsonSchema<T, B> {
     private boolean ignored;
     private boolean preserveUnknownFields;
     private String description;
-    private Class schemaFrom;
+    private TypeRef schemaFrom;
 
     private PropertyOrAccessor(Collection<AnnotationRef> annotations, String name, String propertyName, boolean isMethod) {
       this.annotations = annotations;
@@ -233,9 +233,15 @@ public abstract class AbstractJsonSchema<T, B> {
             preserveUnknownFields = true;
             break;
           case ANNOTATION_SCHEMA_FROM:
-            final Class extractedType = (Class) a.getParameters().get("type");
-            if (extractedType != null) {
-              schemaFrom = extractedType;
+            Object type = a.getParameters().get("type");
+            if (type != null) {
+              if (type instanceof ClassRef) {
+                schemaFrom = (ClassRef) type;
+              } else if (type instanceof Class) {
+                schemaFrom = Types.typeDefFrom((Class) type).toReference();
+              } else {
+                throw new IllegalArgumentException("Unmanaged type passed to the SchemaFrom annotation " + type);
+              }
             }
             break;
         }
@@ -270,7 +276,7 @@ public abstract class AbstractJsonSchema<T, B> {
       return description != null;
     }
 
-    public Class getSchemaFrom() {
+    public TypeRef getSchemaFrom() {
       return schemaFrom;
     }
 
@@ -350,7 +356,7 @@ public abstract class AbstractJsonSchema<T, B> {
         }
 
         if (p.contributeSchemaFrom()) {
-          schemaFrom = Types.typeDefFrom(p.getSchemaFrom()).toReference();
+          schemaFrom = p.getSchemaFrom();
         }
       });
 

